### PR TITLE
New Packages: sequoia cli utilities

### PR DESCRIPTION
**Pull request by jcgruenhage**
#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me
- [ ] I generally don't use the affected packages but briefly tested this PR

### TODO
- [x] sq
- [x] sqop
- [x] sqv
- [x] add man pages
- [x] figure out cross compiling (was broken due to nettle-dev being required on the host as well as the target)
- ~~add shell completions~~ will be done after https://gitlab.com/sequoia-pgp/sequoia/-/issues/798 is fixed.
- [x] tell upstream about adding git tags for openpgp-card-tools ([upstream issue](https://gitlab.com/hkos/openpgp-card/-/issues/32))
- ~~figure out building for armv7l (some lib stuff missing, see [log](https://web.archive.org/web/20220409160215/https://pipelines.actions.githubusercontent.com/serviceHosts/8d180b0e-8647-44ca-8431-6b9f11113866/_apis/pipelines/1/runs/49556/signedlogcontent/6?urlExpires=2022-04-09T16%3A02%3A59.1606901Z&urlSigningMethod=HMACV1&urlSignature=gkM%2FhddrZEYzNMnQZqSFJQcD3ARwMfWvTQLNQA30A9g%3D))~~ marked as nocross for armv7l for now, considering neither we nor upstream really know *why* it's failing to cross compile
- [x] figure out building for x86_64-musl (some deprecated musl specific rust code, see [log](https://web.archive.org/web/20220409160239/https://pipelines.actions.githubusercontent.com/serviceHosts/8d180b0e-8647-44ca-8431-6b9f11113866/_apis/pipelines/1/runs/49556/signedlogcontent/8?urlExpires=2022-04-09T16%3A03%3A21.7631273Z&urlSigningMethod=HMACV1&urlSignature=iRYCydXZC%2FxRhVp8pVRbPgkAk8puf85yX5uMTGmzcGM%3D))

### CODE REVIEW

File: srcpkgs/openpgp-card-tools/template [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+short_desc="Tools for inspecting, configuring and using OpenPGP cards"
+maintainer="Jan Christian Grünhage <jan.christian@gruenhage.xyz>"
+license="MIT, Apache-2.0"
+homepage="https://crates.io/crates/openpgp-card-tools"
+distfiles="https://static.crates.io/crates/${pkgname}/${pkgname}-${version}.crate"
+checksum=ccdd462f47fc92c0e09de78fb0a2ef5be818e836ab7c27b1fdbd54c812008d3d
+
+post_install() {
+	# Upstream doesn't include those files in the crate tarballs,
+	# and doesn't have git tags right now
```
Comment (classabbyamp) [inline srcpkgs/openpgp-card-tools/template:17]: has a request to upstream for this been made?

Comment (jcgruenhage) [reply srcpkgs/openpgp-card-tools/template:17]: Nope, not yet. One of the reasons this PR is still marked as a draft

Comment (jcgruenhage) [reply srcpkgs/openpgp-card-tools/template:17]: Updated the `TODO` section in the PR description, to include this. Resolving thread.

**Comment by jcgruenhage** [comment]
So, I've done some digging, and I'd like to just outright disable armv7l and musl targets for now.

In the case of musl, this is part of time64, the migration to using 64 bit integers for time on 32 bit systems. Rust added a deprecation warning for that, and this warning is currently breaking the build here. There's stuff happening in that area, but it's a bigger problem that's not really related to sequoia directly, so not really something that can be addressed in this PR.

As for armv7l: A crypto library used by sequoia, nettle, for some reason fails to have it's rust bindings compiled for armv7l, because `cross-armv7l-linux-gnueabihf-libc` doesn't provide `gnu/stubs-soft.h`, which is referenced by `gnu/stubs.h`, based on some ifdef magic. I don't get where this is going wrong, but I don't understand enough of the arkane ifdef and crosscompiling magic to dig through this. If anyone else wants to, I'd be more than happy, but I won't invest any more time in trying to figure out why `stubs-soft.h` would even be considered required on a `hf` target.

**Comment by ericonr** [comment]
> In the case of musl, this is part of time64, the migration to using 64 bit integers for time on 32 bit systems. Rust added a deprecation warning for that, and this warning is currently breaking the build here. There's stuff happening in that area, but it's a bigger problem that's not really related to sequoia directly, so not really something that can be addressed in this PR.

That's on Rust, they shouldn't have deprecated the type on 64-bit musl. And more importantly, the software should still be buildable with warnings like that....

> As for armv7l: A crypto library used by sequoia, nettle, for some reason fails to have it's rust bindings compiled for armv7l, because cross-armv7l-linux-gnueabihf-libc doesn't provide gnu/stubs-soft.h, which is referenced by gnu/stubs.h, based on some ifdef magic. I don't get where this is going wrong, but I don't understand enough of the arkane ifdef and crosscompiling magic to dig through this. If anyone else wants to, I'd be more than happy, but I won't invest any more time in trying to figure out why stubs-soft.h would even be considered required on a hf target.

That sounds like `bindgen` not receiving all the flags it should have and looking at the wrong headers/missing some target definitions (probably the latter). We might have a fix for this by using a patched bindgen, iirc.

**Comment by classabbyamp** [comment]
has this branch been rebased recently (after mid-March)? I think i was getting some similar errors with bindgen + cross and solved it by adding this to the rust build helper:
https://github.com/void-linux/void-packages/blob/de7bd4b5077aae0df2e57d7ce43f6feeb08f3f54/common/build-helper/rust.sh#L27-L28

**Comment by jcgruenhage** [comment]
I've looked for quite a while now, and I just can't find it. The `openpgp` crate which fails doesn't have `deny(warnings)` set, I can't find anything adding `-D warnings` to `RUSTFLAGS` either. From the error message in the build, it seems to be the latter, but I can't find where if at all it happens.

File: srcpkgs/sequoia-sop/template [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+short_desc="Implementation of the Stateless OpenPGP CLI using Sequoia"
+maintainer="Jan Christian Grünhage <jan.christian@gruenhage.xyz>"
+license="GPL-2.0-or-later"
+homepage="https://gitlab.com/sequoia-pgp/sequoia-sop/"
+distfiles="https://gitlab.com/sequoia-pgp/sequoia-sop/-/archive/v${version}/sequoia-sop-v${version}.tar.gz"
+checksum=5094965da3fb8c67b7224534ddcd94cf3409d4c7c74e53473d7804f2a343e2aa
+
+post_install() {
+	for page in $(ls man-sqop); do
+		vman man-sqop/${page}
```
Comment (classabbyamp) [inline srcpkgs/sequoia-sop/template:19]:
```suggestion
	for page in man-sqop/*; do
		vman ${page}
```
perhaps (and repeat elsewhere)

File: srcpkgs/sequoia-sq/template [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+distfiles="https://gitlab.com/sequoia-pgp/sequoia/-/archive/sq/v${version}/sequoia-sq-v${version}.tar.gz"
+checksum=e943528be6af2b14c5b9d3a011335233d985ee69f43700d29d99908359f8c345
+conflicts="squirrel"
+
+case "$XBPS_TARGET_MACHINE" in
+	armv7l) nocross="Problems with bindgen and ifdefs trying to import stubs-soft.h on a hf arch";;
+esac
+
+post_patch() {
+	vsed -i "/-Dwarnings/d" "${wrksrc}/.cargo/config.toml"
```
Comment (classabbyamp) [inline srcpkgs/sequoia-sq/template:24]: should be vsed -e expr -i file iirc

**Comment by jcgruenhage** [comment]
TL;DW of the problem here: `sq` depends on nettle-sys for both host (in build.rs) as well as target (for the application itself). Getting `bindgen` to work correctly for both host and target in our cross compilation setup is *super* tricky, and I've sunken many, many hours into it without getting anywhere so far. For most architectures it works anyway, but on armv*l it breaks due to some import trying to get in `/usr/include/gnu/stubs-soft.h` even though they are hard-float architectures, meaning that the file just doesn't exist.

I've tried patching out the requirement for nettle-sys in sq's build.rs script, but that ended up being a way bigger challenge than anticipated, and I'd prefer to just mark this as nocross for the affected architectures for now. If someone could validate that it works on those architectures at all, I'd be very happy about it as I currently don't have any armv7l device anymore to test on, but I suspect that it just works on those.

**Comment by jcgruenhage** [comment]
(Last but not least: The probability that there is even a single user on older arm architectures that wants to use `sq` on Void is very slim, so.. yeah. I don't have any more time to sink into this)

File: srcpkgs/sequoia-sq/template [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+short_desc="Command-line frontend for Sequoia, a new OpenPGP implementation"
+maintainer="Jan Christian Grünhage <jan.christian@gruenhage.xyz>"
+license="GPL-2.0-or-later"
+homepage="https://sequoia-pgp.org/projects/#sq"
+distfiles="https://gitlab.com/sequoia-pgp/sequoia/-/archive/sq/v${version}/sequoia-sq-v${version}.tar.gz"
+checksum=e943528be6af2b14c5b9d3a011335233d985ee69f43700d29d99908359f8c345
+conflicts="squirrel"
+
+case "$XBPS_TARGET_MACHINE" in
+	armv*l) nocross="Requires C libs included in build.rs, which is currently broken in xbps-src. These failures only manifest on a hf archs right now";;
```
Comment (ericonr) [inline srcpkgs/sequoia-sq/template:20]: We talked on IRC about fixing this by updating the bindgen crate, right?

Comment (jcgruenhage) [reply srcpkgs/sequoia-sq/template:20]: Yeah. I tried it with `EXTRA_CLANG_ARGS_<TARGET>` after updating bindgen to 0.59, which I had hoped would help, but without success. The problems persisted

Comment (ahesford) [reply srcpkgs/sequoia-sq/template:20]: I tested a native build on armv7l and it seems to work, so `nocross` seems like the right choice here.

**Pull request merged**